platform/src/subscribe.rs

```diff
@@ -0,0 +1,157 @@
+use crate::syscall_scope::ShareList;
+use crate::Syscalls;
+
+// -----------------------------------------------------------------------------
+// `Subscribe` struct
+// -----------------------------------------------------------------------------
+
+/// A `Subscribe` instance allows safe code to call Tock's Subscribe system
+/// call, by guaranteeing the upcall will be cleaned up before 'scope ends. It
+/// is generally used with the `syscall_scope` function, which offers a safe
+/// interface for constructing `Subscribe` instances.
+pub struct Subscribe<'scope, S: Syscalls, const DRIVER_NUM: u32, const SUBSCRIBE_NUM: u32> {
+    _syscalls: core::marker::PhantomData<S>,
+
+    // Make this struct invariant with respect to the 'scope lifetime.
+    //
+    // Covariance would be unsound, as that would allow code with a
+    // `Subscribe<'static, ...>` to register an upcall that lasts for a shorter
+    // lifetime, resulting in use-after-free if the upcall is invoked.
+    // Contravariance would be sound, but is not necessary and may be confusing.
+    //
+    // Additionally, we want to have at least one private member of this struct
+    // so that code outside this module cannot construct a `Subscribe` without
+    // calling `ShareList::new`.
+    _scope: core::marker::PhantomData<core::cell::Cell<&'scope ()>>,
+}
+
+impl<'scope, S: Syscalls, const DRIVER_NUM: u32, const SUBSCRIBE_NUM: u32> Drop
+    for Subscribe<'scope, S, DRIVER_NUM, SUBSCRIBE_NUM>
+{
+    fn drop(&mut self) {
+        S::unsubscribe(DRIVER_NUM, SUBSCRIBE_NUM);
+    }
+}
+
+impl<'scope, S: Syscalls, const DRIVER_NUM: u32, const SUBSCRIBE_NUM: u32> ShareList<'scope>
+    for Subscribe<'scope, S, DRIVER_NUM, SUBSCRIBE_NUM>
+{
+    // Safety: The safety invariant is inherited from the ShareList trait's
+    // definition. The caller must guarantee that Drop::drop is called on this
+    // Subscribe before the 'scope lifetime ends.
+    unsafe fn new() -> Subscribe<'scope, S, DRIVER_NUM, SUBSCRIBE_NUM> {
+        Subscribe {
+            _syscalls: core::marker::PhantomData,
+            _scope: core::marker::PhantomData,
+        }
+    }
+}
+
+// -----------------------------------------------------------------------------
+// `Upcall` trait
+// -----------------------------------------------------------------------------
+
+/// A Tock kernel upcall. Upcalls are registered using the Subscribe system
```

Comment:
Links to existing docs on upcalls would be appreciated

Reply:
The only thing that comes to mind is [TRD 104](https://github.com/tock/tock/blob/master/doc/reference/trd104-syscalls.md), but that documents the entire syscall interface. I should probably link to that from *somewhere* in `libtock_platform`, but it's unclear where.

Perhaps I should link to TRD 104 from the doc comment on the `Syscalls` trait? I'm not sure it really fits there, as TRD 104 is pretty low level while `Syscalls` is supposed to be higher level. ... I'm also secretly planning to deprecate and delete the `Syscalls` trait entirely in a future refactoring.